hal/mbed_ticker_api.c

```diff
@@ -13,61 +13,178 @@
+    ticker->queue->timestamp = 0;
+    ticker->queue->initialized = true;
+    
+    update_current_timestamp(ticker);
+    update_interrupt(ticker);
+}
+
+/**
+ * Set the event handler function of a ticker instance. 
+ */
+static void set_handler(const ticker_data_t *const ticker, ticker_event_handler handler) { 
+    ticker->queue->event_handler = handler;
+}
+
+/*
+ * Convert a low res timestamp to a high res timestamp. An high resolution 
+ * timestamp is used as the reference point to convert the low res timestamp 
+ * into an high res one. 
+ * 
+ * It is important to note that the result will **never** be in the past. If the 
+ * value of the low res timetamp is less than the low res part of the reference 
+ * timestamp then an overflow is 
+ * 
+ * @param ref: The timestamp of reference. 
+ * @param relative_timestamp: The timestamp to convert. 
+ */
+static us_timestamp_t convert_relative_timestamp(us_timestamp_t ref, timestamp_t relative_timestamp) { 
+     bool overflow = relative_timestamp < ((timestamp_t) ref) ? true : false;
+
+    us_timestamp_t result = (ref & ~((us_timestamp_t)UINT32_MAX)) | relative_timestamp;
+    if (overflow) { 
+        result += (1ULL<<32);
+    }
+
+    return result;
+}
+
+/**
+ * update the current timestamp value of a ticker.
+ */
+static void update_current_timestamp(const ticker_data_t *const ticker) { 
+    ticker->queue->timestamp = convert_relative_timestamp(
+        ticker->queue->timestamp, 
+        ticker->interface->read()
+    );
+}
+
+/**
+ * update the interrupt with the appropriate timestamp. 
+ * if there is no interrupt scheduled or the next event to execute is in more 
+ * than MBED_TICKER_INTERRUPT_TIMESTAMP_MAX_DELTA us from now then the 
+ * interrupt will be set to MBED_TICKER_INTERRUPT_TIMESTAMP_MAX_DELTA us from now. 
+ * Otherwise the interrupt will be set to head->timestamp - queue->timestamp us.
+ */
+static void update_interrupt(const ticker_data_t *const ticker) { 
+    update_current_timestamp(ticker);
+    uint32_t diff = MBED_TICKER_INTERRUPT_TIMESTAMP_MAX_DELTA;
+
+    if (ticker->queue->head) { 
+        diff = MBED_MIN(
+            (ticker->queue->head->timestamp - ticker->queue->timestamp), 
+            MBED_TICKER_INTERRUPT_TIMESTAMP_MAX_DELTA
+        );
+    } 
+
+    ticker->interface->set_interrupt(
+        ticker->queue->timestamp + diff
+    );
+}
+
+void ticker_set_handler(const ticker_data_t *const ticker, ticker_event_handler handler) {
+    initialize(ticker);
+    set_handler(ticker, handler);
 }
 
-void ticker_irq_handler(const ticker_data_t *const data) {
-    data->interface->clear_interrupt();
+void ticker_irq_handler(const ticker_data_t *const ticker) {
+    ticker->interface->clear_interrupt();
 
     /* Go through all the pending TimerEvents */
     while (1) {
-        if (data->queue->head == NULL) {
-            // There are no more TimerEvents left, so disable matches.
-            data->interface->disable_interrupt();
```

Comment:
>  The workaround is to track overflow at this layer but it requires to have interrupts always on.

as in - if none, schedule max delta as it is in update_interrupts, and count overflows here in higher layer. I got it now